### DISCUSSION TITLE
Filled Color Update Question

### DISCUSSION THREAD

**Discussion by xiaor2** [visgl/deck.gl#7579] [Q&A] [answered]
```const overlay = new DeckOverlay({
      layers: [new GeoJsonLayer({
        id: "geojson-layer",
        data: my_pollutant,
        pickable: true,
        stroked: true,
        filled: true,
        extruded: true,
        lineWidthUnits: "pixels",
        getFillColor: d => [255, 255 * d.properties.TotalPM25, 0, 80],
        getLineColor: [255, 0, 0, 20],
        getPointRadius: 100,
        getLineWidth: 0,
        onClick: async (info, event) => {
          const SOA = await z.getItem("SOA").then(async data => await data.get([0, info.index, slice(null, 52411)])).then(async data => await data.data)
          let id = 0;
          for (let s of SOA) {
            my_pollutant.features[id].properties.TotalPM25 = SOA[id] * 10 ** 14
            id = id + 1;
          }
        },
      })],
     ;
```

Hi, I tried to use the onClick() method to change the data and the color. The data is changed correctly, but the filled color is not changed at all. I will be grateful if you can give me some suggestions.

**Comment by lianzhao** [comment, marked as answer]
Try [updateTriggers](https://deck.gl/docs/api-reference/core/layer#updatetriggers)
```js
new GeoJsonLayer({
  // ...
  updateTriggers: {
      getFillColor: [setThisObjectWhenYouWantToUpdateYourFillColor]
  }
})
```

**Reply by xiaor2** [reply]
Thanks for your suggestion! The TotalPM25 changed as I updated the data. I think I should put d in the square brackets, but obviously, d is not defined, it is just a name I used for the arrow function. I wonder what should I put in there?

**Reply by lianzhao** [reply]
Well, I think the most simple solution is updating your `data` prop (update it's reference, not it's content) in `onClick` callback. But this is not the recommend way. Please refer to https://deck.gl/docs/developer-guide/performance

I think it's better to
```js
let pm25Map = new Map<string, number>(); // feature id -> pm25 value
const layer = new GeoJsonLayer({
  // ...
  getFillColor: d => [255, 255 * pm25Map.get(d.properties.id), 0, 80], // assuming you have an 'id' prop
  updateTriggers: {
    getFillColor: [pm25Map],
  },
  onClick: () => {
    pm25Map = ...; // update pm25Map reference
  },
});
```

**Reply by xiaor2** [reply]
Thanks! I think the code below follows the recommended way.
All the PM25 data in ```my_pollutant``` are 0, and all the PM25 data in ```pollutant``` are not 0.
In my onclick, I made a new map to replace the original map and set the ```updateTriggers```. When I cilcked on this map, the getFillColor is still not called. Did I write something wrong?

```

let PM25Map = new Map();

for (let p of my_pollutant.features) {
  PM25Map.set(p.id, 0);
}

// ...

const overlay = new DeckOverlay({
  layers: [new GeoJsonLayer({
          id: 'pollutant',
          data: my_pollutant,
          pickable: true,
          stroked: true,
          filled: true,
          extruded: true,
          lineWidthUnits: 'pixels',
          getFillColor: (data) => {
            console.log("filled")
            return [255, 255 * PM25Map.get(data.id), 0, 80]},
          getLineColor: [255, 0, 0, 20],
          getPointRadius: 100,
          getLineWidth: 0,
          onClick: (info, event) => {
            PM25Map = new Map()
            for (let p of pollutant.features) {
              PM25Map.set(p.id, p.properties.TotalPM25)
            }
            console.log("It's done!")
          },
          updateTriggers: {
            all: [PM25Map]
          },
        })]

```

**Reply by Pessimistress** [reply]
You need to call `overlay.setProps({layers: [...]})` when you update your `updateTriggers`. When your `GeoJsonLayer` is constructed, `props.updateTriggers` is set to the value of `PM25Map` at the time.

A equivalent of your problem in the code above is that you are writing

```js
let a = 0;
const b = a + 1;
a = 1;
```
And asking why `b` doesn't change.

**Reply by xiaor2** [reply]
It works now! I called the ```overlay.setProps({layers: [...]})``` to construct a new layer with different id and it renders. Thank you！
